src/compiler/mod.rs

```diff
@@ -100,31 +130,86 @@ impl<'a> Compiler<'a> {
         }
     }
 
+    fn push_to_datatable(&mut self, value: PactType<'a>) -> Result<(), CompileErr> {
+        self.data_table.push(value);
+        if self.data_table.len() > MAX_ENTRIES {
+            Err(CompileErr::DataTableFull)
+        } else {
+            Ok(())
+        }
+    }
+
     /// Compile an assertion AST node
     fn compile_assertion(&mut self, assertion: &'a ast::Assertion) -> Result<(), CompileErr> {
-        let comparator_op = compile_comparator(&assertion.2)?;
-        self.bytecode.push(comparator_op.into());
-
-        let lhs_load_op = self.compile_subject(&assertion.0)?;
-        match lhs_load_op {
-            OpCode::LD_INPUT(index) | OpCode::LD_USER(index) => {
-                self.bytecode.push(lhs_load_op.into());
-                self.bytecode.push(index);
-            }
-            _ => panic!("unreachable"),
+        let lhs_load = self.compile_subject(&assertion.0)?;
+        let (comparator_op, invert) = compile_comparator(&assertion.1, &assertion.2)?;
+        let rhs_load = self.compile_subject(&assertion.3)?;
+
+        // Determine the Load Order
+        let (load, flip) = match lhs_load.1 {
+            LoadSource::Input => match rhs_load.1 {
+                LoadSource::Input => (OpLoad::INPUT_VS_INPUT, false),
+                LoadSource::DataTable => (OpLoad::INPUT_VS_USER, false),
+            },
+            LoadSource::DataTable => match rhs_load.1 {
+                LoadSource::Input => (OpLoad::INPUT_VS_USER, true),
+                _ => return Err(CompileErr::InvalidCompare),
+            },
         };
 
-        // TODO: Imperative is ignored for now. In future it will set/flip the truthiness of the comparator op
+        // Check whether we need to flip the load indices
+        // to meet the load order
+        let indices = if flip {
+            OpIndices {
+                lhs: rhs_load.0,
+                rhs: lhs_load.0,
+            }
+        } else {
+            OpIndices {
+                lhs: lhs_load.0,
+                rhs: rhs_load.0,
+            }
+        };
 
-        let rhs_load_op = self.compile_subject(&assertion.3)?;
-        match rhs_load_op {
-            OpCode::LD_INPUT(index) | OpCode::LD_USER(index) => {
-                self.bytecode.push(rhs_load_op.into());
-                self.bytecode.push(index);
+        // A flip means we need to redefine inequalities
+        let (comparator_op, invert) = if flip {
```

Comment:
Suggestion: move this to its own utility function. especially since this is used twice when we need to invert the inequality twice.

Reply:
the two inversions look smilar but is really serves different purposes. We can disregard my comment above